frontend/model/state.js

```diff
@@ -99,36 +170,52 @@ sbp('sbp/selectors/register', {
-        }).reduce((contractID: string | boolean, member: Object) => {
-          if (contractID) return contractID
-          if (member.status === PROFILE_STATUS.ACTIVE && member.joinedHeight == null) {
-            return groupID
-          }
-          return false
-        }, false)
-      }).forEach((contractID) => {
-        if (!contractID) return
-        sbp('gi.actions/group/upgradeFrom1.0.7', { contractID }).catch(e => {
-          console.error('[state/vuex/postUpgradeVerification] Error during gi.actions/group/upgradeFrom1.0.7', contractID, e)
+      Object.entries(state[ourIdentityContractId].groups)
+        .filter(([groupID, { hasLeft }]: [string, Object]) => {
+          return !hasLeft &&
+           state[groupID]?.chatRooms &&
+           (!Array.isArray(contractIDHints) || contractIDHints.includes(groupID))
         })
-      })
-    })();
-    (() => {
+        .map(([groupID]) => {
+          // $FlowFixMe[incompatible-use]
+          const chatRooms = state[groupID].chatRooms
+          const needsUpgrade = ((Object.values(chatRooms): any): Object[])
+            .flatMap(({ members }): Object => Object.values(members))
+            .some(member =>
+              member.status === PROFILE_STATUS.ACTIVE && member.joinedHeight == null
+            )
+
+          return needsUpgrade ? groupID : null
+        })
+        .filter(Boolean)
+        .forEach((contractID) => {
+          if (!contractID) return
+          sbp('gi.actions/group/upgradeFrom1.0.7', { contractID }).catch(e => {
+            console.error('[state/vuex/postUpgradeVerification] Error during gi.actions/group/upgradeFrom1.0.7', contractID, e)
+          })
+        })
+    }, 'gi.contracts/group')
+    contractUpdate(state, (contractIDHints: ?string[]) => {
       // Upgrade from version 1.0.8 to a newer version
       // The new chatroom contracts have an admin IDs list
       // This code checks if the attribute is missing, and if so, issues the
       // corresponing upgrade action.
-      const needsUpgrade = (chatroomID) => !Array.isArray(state[chatroomID]?.attributes?.adminIDs)
+      const needsUpgrade = (chatroomID) => {
+        // Restrict updates to recently added contracts
+        if (Array.isArray(contractIDHints) && !contractIDHints.includes(chatroomID)) return false
+        return !!state[chatroomID]?.attributes && !Array.isArray(state[chatroomID].attributes.adminIDs)
+      }
 
       const upgradeAction = async (contractID: string, data?: Object) => {
         try {
           await sbp('gi.actions/chatroom/upgradeFrom1.0.8', { contractID, data })
         } catch (e) {
+          // If the action failed because the upgrade has already happened, we
+          // can safely ignore the error
+          if (e.message?.includes('Upgrade can only be done once')) {
+            console.warn(`[state/vuex/postUpgradeVerification] Error during gi.actions/chatroom/upgradeFrom1.0.8 for ${contractID}:`, e)
+            return
+          }
```

Comment:
Why issue the action at all just to capture this error? Is it possible to instead avoid sending the action?

Reply:
I don't think so because new events could have been received between the time the check was made and the time the action is sent.